src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectFactory.cs

```diff
@@ -99,6 +99,8 @@ public VisualStudioProject CreateAndAddToWorkspace(string projectSystemName, str
                 {
                     w.OnProjectAdded(projectInfo);
                 }
+
+                _visualStudioWorkspaceImpl.RefreshProjectExistsUIContextForLanguage(language);
```

Comment:
Should this be done at the end of the function just before returning the project (so that the properties being set below will definitely be avaialble to anyone querying projects in the event handler)? Or is that already guaranteed some other way?

Reply:
There is absolutely no guarantee that the projects will be *ready* with the UIContext; even if I move this they still won't have files or references for example.

I initially liked this idea but then it resulted in a fairly long comment explaining why that wasn't adding a race because I'm moving the setting of this state outside of the lock that's being acquired. (Answer: the refresh takes it's own lock, but the project can't be removed since it hasnt been handed out yet.) It's safe but this isn't really saving you much -- at most a few workspace change events when you subscribe, and only then on the first project. If those workspace change events were a problem, we should address them directly.

Reply:
Ah indeed - makes sense that consumer would have to rely on workspace change events as opposed to this.